src/sim/kernel_workload.cc

```diff
@@ -104,12 +104,28 @@ KernelWorkload::initState()
     };
     if (params().object_file != "")  {
         if (params().addr_check) {
+            auto memory_size = system->memSize();
+            auto kernel_size = _end - _start;
             // Validate kernel mapping before loading binary
-            fatal_if(!system->isMemAddr(mapper(_start)) ||
+            if (kernel_size > memory_size){
+                fatal_if(!system->isMemAddr(mapper(_start)) ||
                     !system->isMemAddr(mapper(_end)),
```

Comment:
I think there is a misunderstanding. The kernel_size > memory size should be the fatal_if condition. No need to have an extra if-else. So the code should be something like:

fatal_if(kernel_size > memory_size, [...]);
fatal_if(system->isMemAddr[...]);